### PR TITLE
GUI-2252 converted to client-side filtering, except for aws

### PR DESCRIPTION
**Pull request by dkavanagh**
converted to client-side filtering, except for aws

https://eucalyptus.atlassian.net/browse/GUI-2252

### CODE REVIEW

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com/job/user-console-PR-monitor/1209/
Test PASSed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com/job/user-console-PR-monitor/1210/
Test PASSed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com/job/user-console-PR-monitor/1212/
Test PASSed.

File: eucaconsole/static/js/widgets/image_picker.js [hunk 1]
```diff
@@ -81,32 +87,132 @@ angular.module('ImagePicker', ['EucaConsoleUtils', 'MagicSearch'])
...
+                            return true;
+                        }
+                    }
+                };
+                var filterFunc = function(item) {
+                    // handle special case of empty facet value, match all
+                    if (facets[key].indexOf("") > -1) {
+                        return true;
+                    }
+                    var val = item.hasOwnProperty(key) && item[key];
```
Comment (coyote240) [inline eucaconsole/static/js/widgets/image_picker.js]: No need for the hasOwnProperty guard as best I can tell, unless you're checking that an assignment was done on this object and not a parent object.  hasOwnProperty is really used for edge-case workarounds to prototypical inheritance, and not often used in general.  item[key] will return undefined if it's not there, which you're testing for.

File: eucaconsole/static/js/widgets/image_picker.js [hunk 2]
```diff
@@ -81,32 +87,132 @@ angular.module('ImagePicker', ['EucaConsoleUtils', 'MagicSearch'])
...
+                angular.forEach(tmp, function(item) {
+                    var facet = item.split('=');
+                    if (this[facet[0]] === undefined) {
+                        this[facet[0]] = [];
+                    }
+                    this[facet[0]].push(facet[1]);
+                }, facets);
+                var results = unfilteredItems;
+                // filter results
+                var matchFunc = function(val) {
```
Comment (coyote240) [inline eucaconsole/static/js/widgets/image_picker.js]: These functions can be defined in the controller function.  Here they will be redefined every time facetFilterItems is called.  I'd move them to the controller and add a second parameter for facets.  Reduces the need for the closure, makes it easier to reason about the surrounding code and forces you to rename the functions so they can be understood by name.

File: eucaconsole/static/js/widgets/image_picker.js [hunk 3]
```diff
@@ -81,32 +87,132 @@ angular.module('ImagePicker', ['EucaConsoleUtils', 'MagicSearch'])
...
+            $scope.items = filteredItems;
         };
         $scope.imageSelected = function (item) {
             $scope.selectedImageParam = item.id;
             $scope.$emit('imageSelected', item);
         };
         $scope.$on('searchUpdated', function($event, query) {
-            $scope.jsonEndpoint = decodeURIComponent($scope.jsonEndpointPrefix + "&" + query);
-            $scope.getItems();
+            // update url
```
Comment (coyote240) [inline eucaconsole/static/js/widgets/image_picker.js]: AngularJS's $location service does much of this for you.  I'd favor this over rolling your own.

File: eucaconsole/static/js/widgets/image_picker.js (hunk 2 above)
Comment (dkavanagh) [inline eucaconsole/static/js/widgets/image_picker.js]: Mostly, these were created to get out of inlining them, but I can see putting them into the controller scope.

File: eucaconsole/static/js/widgets/image_picker.js (hunk 3 above)
Comment (dkavanagh) [inline eucaconsole/static/js/widgets/image_picker.js]: I looked at this. trimmed it down to 2 lines of code that mostly worked, BUT, $location.absUrl() inserts a '#' character between the path and the search parts of the URL (not the end where a fragment ought to go). That screws up magic search (which might be a secondary problem).

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com/job/user-console-PR-monitor/1217/
Test PASSed.

**Comment by coyote240** [comment]
Thanks for making the updates.  Worth digging into the $location issue more, but not worth holding up the works today.  Merging.

**Pull request merged**